Simplify and document Python and Perl requirements for running all our converters

magnumripper opened the pull request:
Many *2john tools just work, but following this virtually all of them work with little effort. This is probably also great for distro package maintainers.

Also a few bug fixes found while at it.

magnumripper committed:
Document Python modules installation

Amend requirements.txt and move it to run directory, as previously
discussed in #5475.  Add everything needed for being able to run all
of our python tools, including instructions in doc/README-UBUNTU.

All modules exist as Ubuntu packages except protobuf, pyhanko and
pytsk3, so add instructions for that as well as and alternative.

Also fixes permissions on a few python scripts.

magnumripper committed:
Change shebang for pse2john

It requires python2, because the 'pysap' module is no longer maintained.

Apparently there is a fork of pysap for Python3 at github.com/Esri/pysap
though.

magnumripper committed:
Document Perl libs installation

Add a 'cpanfile' for easy installation of everything needed for running
the various converters.

Add a separate 'cpanfile.pass_gen' for pass_gen.pl as there's a lot of
dependencies and that tool is not a converter.

Add instructions to README-UBUNTU.

magnumripper committed:
Perl converter bugfixes

Do not use a legacy codepage's copyright sign. In fact, do not even
use UTF-8 for it but (c), as per Solar's wish long ago IIRC.

leet.pl: We can't send perl options via env, so need to change
"/usr/bin/env perl -l" to just "/usr/bin/perl -l" (the -l doesn't
have any trivial "use xx" alternative).

Other minor syntax fixes probably not needed for older perl versions.

magnumripper committed:
apop2john.py: Python3 fixes

Some redundant escapes produced warnings such as "SyntaxWarning: "\+"
is an invalid escape sequence. Such sequences will not work in the
future".

exploide commented on doc/INSTALL-UBUNTU:
```diff
@@ -132,3 +132,17 @@ build CPU fallback chains for any x86-64 CPU.
+    sudo apt-get install python3 python3-venv python3-pip
+    python3 -m venv venv
+    source venv/bin/activate
+    pip3 install `cat requirements.txt`
```
`pip3 install -r requirements.txt`
No need for command substitution.

magnumripper replied:
Thanks, fixed. I also replaced `source` with `.` as that is more portable.

magnumripper commented on run/rexgen2rules.pl:
```diff
@@ -74,7 +74,7 @@ sub check_parity {
-            if (/\\g{(\d+)}/ && $1 > $k) {
+            if (/\\g\{(\d+)}/ && $1 > $k) {
                 die "wrong back reference $_"
             }
```
@AlekseyCherepanov Is this correct? Current Perl complains about the old line but I'm not entirely sure I understand the use of curly brackets here.

AlekseyCherepanov replied:
Yes, the fix is right. It might be more stylish to escape both curly brackets with backslashes. OTOH `$token_re` has `\{...}` too (`my $token_re = qr/(\\g\{\s*\d+\s*}|...`). There are backslashes for both braces in other places.

The regular expression in the fix matches substrings like `\g{123}`. The braces were intended to be literal. Braces are magic/active chars in Perl's regular expressions to denote repetition. It was my mistake to forget backslashes there. The magic did not work with `(\d+)` inside and older Perl was ok with that. Newer Perl is more strict parsing this nonsense. Thank you for finding and fixing it!

AlekseyCherepanov commented on run/apop2john.py:
The first argument of `re.search` function is a string containing regular expression. While there are excessive backslashes, `+` at the beginning should be escaped to be literal. Otherwise there would be an exception:
```python3
>>> re.search(b'+OK', b'+OK')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.11/re/__init__.py", line 176, in search
[...]
    raise source.error("nothing to repeat",
re.error: nothing to repeat at position 0
```
I think `\+OK` should be kept with its backslash. Other backslashes seem unneeded indeed.

AlekseyCherepanov replied:
`b'\+OK ...'` is the same as `b'\\+OK ...'`. But [howto](https://docs.python.org/3/howto/regex.html#the-backslash-plague) says that the first variant yields a `DeprecationWarning` in the newest Python. So this backslash should be doubled.

magnumripper replied:
Thanks! That is insanely backwards... until you read the howto. I'll fix it.

magnumripper replied:
I guess we could also use raw bytes, as in `re.search(br'\+OK .* <.+>', pkt)`, avoiding one layer of escapes. That is easier to read.

magnumripper committed:
office2john: Drop the hardcoded (outdated) copy of 'olefile' module

Just import it instead. The module was added to requirements.txt and
as a simpler alternative, INSTALL-UBUNTU now includes the needed
Ubuntu packages for it.

magnumripper commented:
I now added a commit that drops the hardcoded and outdated copy of 'olefile' module from office2john.py, in favor of a normal import. Also added instructions to INSTALL-UBUNTU on how to install nearly all required modules as Ubuntu packages.

magnumripper commented:
> Also added instructions to INSTALL-UBUNTU on how to install nearly all required modules as Ubuntu packages.

I will do this for the needed Perl libs as well.

Edit: I dropped that idea, there are many packages needed that doesn't exist as Ubuntu packages.

Edit 2: I ended up adding the few available, for easier use of eg. `7z2john.pl` without hassling with CPAN.

magnumripper committed:
Update doc/INSTALL and INSTALL-FEDORA similar to INSTALL-UBUNTU

magnumripper commented:
This should be ready to merge now.

magnumripper committed:
Add bsddb3 to Python requirements, for bitcoin2john.py

magnumripper committed:
Add python-is-python3 to INSTALL-UBUNTU's list of packages

See #5786
